### PR TITLE
Implement Monaco editor for Alertmanager YAML editor

### PR DESCRIPTION
**Pull request by dtaylor113**
### Before

image

### After
image

- Added prop to show/hide the Reload button
- Added prop to pass in save function
- Updated Integration tests

### CODE REVIEW

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx [hunk 1]
```diff
@@ -28,70 +27,59 @@ const AlertManagerYAMLEditor: React.FC<AlertManagerYAMLEditorProps> = ({
...
-    <div className="co-m-pane__body">
-      <form
-        className="co-m-pane__body-group co-alert-manager-yaml-form co-m-pane__form"
-        onSubmit={save}
-      >
-        <p className="co-alert-manager-yaml__explanation">
+    <>
+      <div className="co-alertmanager-yaml-editor__header">
+        <p className="help-block">
           Update this YAML to configure Routes, Receivers, Groupings and other Alertmanager settings
```
Comment (kyoto) [inline frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: Should have `.` at the end of the sentence.

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx [hunk 2]
```diff
@@ -28,70 +27,59 @@ const AlertManagerYAMLEditor: React.FC<AlertManagerYAMLEditorProps> = ({
...
 
   return (
-    <div className="co-m-pane__body">
-      <form
-        className="co-m-pane__body-group co-alert-manager-yaml-form co-m-pane__form"
-        onSubmit={save}
-      >
-        <p className="co-alert-manager-yaml__explanation">
+    <>
+      <div className="co-alertmanager-yaml-editor__header">
```
Comment (kyoto) [inline frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: There are layout issues at narrow screen widths. I think you'd be better off using `co-m-nav-title` like.

```
<>
  <div className="co-m-nav-title">
    <p className="help-block">
      ...
    </p>
    {errorMsg && ...
    {successMsg && ...
  </div>
  <EditAlertmanagerYAML onSave={save} obj={alertManagerYamlStr} />
</>
```

Then `co-alertmanager-yaml-editor__header` and `co-alertmanager-yaml-alert` can be simplified or possibly removed.

**Comment by kyoto** [comment]
Generally LGTM 👍 . Just one issue with the header area layout.

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx (hunk 2 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: Replaced `co-alertmanager-yaml-editor__header` with `co-m-nav-title`.   
`co-alertmanager-yaml-alert` is used to align the inline alerts:
image

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx (hunk 1 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: fixed

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx (hunk 2 above)
Comment (kyoto) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: Looks better, but I'm still seeing layout issues at small screen widths. I think the solution is to put the `Alert`s inside the `co-m-nav-title`, which should handle the alignment automatically.

screenshot

Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: Hi @kyoto, thanks, putting alerts inside of `co-m-nav-title` did fix the left alignment, but still wasn't happy with large bottom gap between alerts and text editor.  Updated `co-alertmanager-yaml-alert` to display properly on **small screen**:
image

**Md/Lg screen:**

image

- thanks

File: frontend/public/components/edit-yaml.jsx [hunk 3]
```diff
@@ -796,12 +796,16 @@ const EditYAML_ = connect(stateToProps)(
                             type="submit"
                             variant="primary"
                             id="save-changes"
-                            onClick={() => this.save()}
+                            onClick={() =>
+                              this.props.onSave
+                                ? this.props.onSave(this.getEditor().getValue())
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]: I'd move this logic into `save()`

File: frontend/public/components/edit-yaml.jsx [hunk 4]
```diff
@@ -666,7 +666,7 @@ const EditYAML_ = connect(stateToProps)(
       });
 
       const { error, success, stale, yaml, height, showSidebar } = this.state;
-      const { obj, download = true, header } = this.props;
+      const { obj, download = true, header, showReload = true } = this.props;
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]: Booleans that default to true are confusing. I'd make this `noReload`.

File: frontend/integration-tests/views/monitoring.view.ts [hunk 5]
```diff
@@ -29,7 +29,7 @@ export const saveButton = $('button[type=submit]');
 export const modalConfirmButton = $('#confirm-action');
 
 // YAML form
-export const alertManagerYamlForm = $('.co-alert-manager-yaml__form');
+export const alertManagerYamlEditor = $('.yaml-editor');
```
Comment (spadgett) [inline frontend/integration-tests/views/monitoring.view.ts]: I'd make this generic and move it into yaml.view.ts

File: frontend/public/components/monitoring/_monitoring.scss [hunk 6]
```diff
@@ -232,18 +232,17 @@ $tooltip-background-color: #151515;
...
-  margin-bottom: 0;
-}
-
-.co-alert-manager-yaml__form-entry-wrapper {
-  position: relative;
-}
-
-.co-alert-manager-yaml__form co-file-dropzone__textarea {
-  min-height: 400px;
+.btn-edit-alert-routing {
```
Comment (spadgett) [inline frontend/public/components/monitoring/_monitoring.scss]: I'm confused by this change. It looks unrelated?

File: frontend/public/components/monitoring/_monitoring.scss [hunk 7]
```diff
@@ -232,18 +232,17 @@ $tooltip-background-color: #151515;
...
 }
 
 .co-alert-manager-config__edit-alert-routing-btn {
   margin-bottom: 10px;
 }
+
+.co-alertmanager-yaml-alert {
+  margin-bottom: 4px;
+  @media (min-width: $screen-sm-min) {
+    margin-bottom: -26px;
```
Comment (spadgett) [inline frontend/public/components/monitoring/_monitoring.scss]: Can you explain why the negative margin is needed?

**Comment by spadgett** [comment]
@alimobrem fyi

**Comment by spadgett** [comment]
We should hide `View shortcuts` since they won't work here. Instead of a `showReload` prop, we might want a general prop to indicate this isn't a k8s resource.

File: frontend/public/components/monitoring/_monitoring.scss (hunk 7 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/_monitoring.scss]: Previously, the Alerts for the YAML Editor appeared between the textarea and buttons:
image

The alert states (error & success) were being handled within the YAML Editor and based on saving a k8s resources.  I'm hijacking the YAML Editor and providing my own `save()` function and handing error & success in a higher component, and thus putting the alerts above the textarea.   If I don't adjust the bottom there is a very large gap, which may be minimized if we remove the `view shortcuts` link:
image

I'll hide the `view shortcuts` link and thus probably won't have to use a negative margin.
I could also try to have my custom `save()` throw exceptions to be caught by the YAML editor and thus displayed below the textarea.

File: frontend/public/components/monitoring/_monitoring.scss (hunk 6 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/_monitoring.scss]: Think this was a merge issue, it's not used and seems to be an older/duplicate of `co-alert-manager-config__edit-alert-routing-btn`  -I'll remove

File: frontend/public/components/monitoring/_monitoring.scss (hunk 7 above)
Comment (spadgett) [reply frontend/public/components/monitoring/_monitoring.scss]: I wonder if we shouldn't pass this as `children` to the edit yaml component so we can show it in the same spot.

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx [hunk 8]
```diff
@@ -28,70 +27,61 @@ const AlertManagerYAMLEditor: React.FC<AlertManagerYAMLEditorProps> = ({
...
-  );
+  const alertManagerYamlStr = !_.isEmpty(encodedAlertManagerYaml)
+    ? Base64.decode(encodedAlertManagerYaml)
+    : '';
 
-  const save = (e) => {
-    e.preventDefault();
-    if (_.isEmpty(alertManagerYamlStr)) {
+  const save = (yaml: string) => {
+    if (_.isEmpty(yaml)) {
```
Comment (spadgett) [inline frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: We should really test that this parses as YAML before patching the config

Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: Is that with `safeLoad()`?  Or will `safeDump()` also throw error if invalid YAML?

Comment (spadgett) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: `safeLoad`. We should catch and display YAML parsing errors.

File: frontend/public/components/monitoring/alert-manager-utils.tsx [hunk 9]
```diff
@@ -36,6 +36,15 @@ export const patchAlertManagerConfig = (
   secret: K8sResourceKind,
   yaml: object | string,
 ): Promise<any> => {
+  if (_.isString(yaml)) {
```
Comment (spadgett) [inline frontend/public/components/monitoring/alert-manager-utils.tsx]: Is this ever not a string?

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx (hunk 2 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: UPDATE:  Passed in high level component's Alerts as children to YAML Editor.  Ie. Alerts now appear below textarea -no need for custom css.

File: frontend/public/components/edit-yaml.jsx (hunk 3 above)
Comment (dtaylor113) [reply frontend/public/components/edit-yaml.jsx]: moved.

File: frontend/public/components/monitoring/alert-manager-utils.tsx (hunk 9 above)
Comment (spadgett) [reply frontend/public/components/monitoring/alert-manager-utils.tsx]: I think this is better inside `save()` instead of here.

File: frontend/public/components/edit-yaml.jsx (hunk 4 above)
Comment (dtaylor113) [reply frontend/public/components/edit-yaml.jsx]: changed to `nonK8s`, but seems a little wonky specifing `{stale && !nonK8s && ( ...`

File: frontend/integration-tests/views/monitoring.view.ts (hunk 5 above)
Comment (dtaylor113) [reply frontend/integration-tests/views/monitoring.view.ts]: moved.

File: frontend/public/components/monitoring/alert-manager-utils.tsx [hunk 10]
```diff
@@ -36,6 +36,15 @@ export const patchAlertManagerConfig = (
...
+    try {
+      safeLoad(yaml);
+    } catch (e) {
+      return new Promise(() => {
+        throw new Error(`Error parsing Alertmanager YAML: ${e}`);
```
Comment (spadgett) [inline frontend/public/components/monitoring/alert-manager-utils.tsx]: Use `Promise.reject`, but I don't think it's necessary if you just make the check inside of `save()`.

File: frontend/public/components/monitoring/_monitoring.scss (hunk 7 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/_monitoring.scss]: Passing the Alerts as children worded great.  -thanks

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx [hunk 11]
```diff
@@ -1,25 +1,24 @@
...
 import { Helmet } from 'react-helmet';
 
-const DroppableFileInput = (props) => (
+const EditAlertmanagerYAML = (props) => (
   <AsyncComponent
-    loader={() => import('../utils/file-input').then((c) => c.DroppableFileInput)}
     {...props}
+    loader={() => import('../edit-yaml').then((c) => c.EditYAML)}
+    create={false}
+    nonK8s={true}
```
Comment (spadgett) [inline frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]:
```suggestion
    nonK8s
```

File: frontend/public/components/edit-yaml.jsx [hunk 12]
```diff
@@ -329,7 +329,14 @@ const EditYAML_ = connect(stateToProps)(
     }
 
     save() {
+      const { create, onSave } = this.props;
       let obj;
+
+      if (!create && onSave) {
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]: Why do we need to check `!create`?

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx (hunk 8 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: Ok, updated `alert-manager-utils.patchAlertManagerConfig` to throw error if `safeLoad` fails.

File: frontend/public/components/monitoring/alert-manager-utils.tsx (hunk 9 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-utils.tsx]: yes, it is an object when called from a few other places.  `patchAlertManagerConfig` is a common helper function.  All `save()`'s use `patchAlertManagerConfig`

File: frontend/public/components/edit-yaml.jsx (hunk 12 above)
Comment (dtaylor113) [reply frontend/public/components/edit-yaml.jsx]: I was specifically passing in a save function, not intending to use for `create`, but I suppose I could not do the create check.

File: frontend/public/components/edit-yaml.jsx (hunk 4 above)
Comment (spadgett) [reply frontend/public/components/edit-yaml.jsx]: > changed to `nonK8s`, but seems a little wonky specifing `{stale && !nonK8s && ( ...`

Is `stale` ever true if this isn't a k8s object? I wonder if the `nonK8s` check is needed

File: frontend/public/components/monitoring/alert-manager-utils.tsx (hunk 9 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-utils.tsx]: yes, it is an object when called from a few other places.  `patchAlertManagerConfig` is a common helper function.  All `save()`'s use `patchAlertManagerConfig`

Comment (spadgett) [reply frontend/public/components/monitoring/alert-manager-utils.tsx]: I'm not clear on this answer. There is more than one `save()`? If this logic moves to `save()` wouldn't it always be a string?

File: frontend/public/components/edit-yaml.jsx (hunk 12 above)
Comment (spadgett) [reply frontend/public/components/edit-yaml.jsx]: For your use, but there's nothing about the prop that is specific to edit vs save. It'd be surprising to have `onSave` and not have it called because another prop is set to a certain value.

File: frontend/public/components/edit-yaml.jsx (hunk 4 above)
Comment (dtaylor113) [reply frontend/public/components/edit-yaml.jsx]: for `stale` the `nonK8s` check isn't necessary.  Still have ` {!create && !nonK8s && (..` - I just don't like '!non...`.

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx (hunk 8 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: fixed

File: frontend/public/components/monitoring/alert-manager-utils.tsx (hunk 9 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-utils.tsx]: fixed.  moved to alertmanager-yaml-editor's `save()`.

File: frontend/public/components/monitoring/alert-manager-yaml-editor.tsx (hunk 11 above)
Comment (dtaylor113) [reply frontend/public/components/monitoring/alert-manager-yaml-editor.tsx]: fixed

File: frontend/public/components/edit-yaml.jsx (hunk 12 above)
Comment (dtaylor113) [reply frontend/public/components/edit-yaml.jsx]: ok, removed `!create` check

File: frontend/public/components/edit-yaml.jsx (hunk 4 above)
Comment (spadgett) [reply frontend/public/components/edit-yaml.jsx]: You could change `nonK8s` to something like `anyYAML` or `genericYAML`. I think that's more clear anyway.

Comment (spadgett) [reply frontend/public/components/edit-yaml.jsx]: We should be skipping all the logic that checks for schemas, etc., when set.

Comment (dtaylor113) [reply frontend/public/components/edit-yaml.jsx]: I like it, changed to `genericYAML` -thanks
